[BE][Team-13][프레디, Starve] 목업 api 배포

Dae-Hwa opened the pull request:
안녕하세요 프레디와 스타브입니다.

작업내용은 아래와 같습니다.

- 샘플 Mockup API 에 있는 데이터 json 파일로 변환하여 관리
  - jack라이브러리 직접 이용하여 json파일 파싱하여 사용
- 기본적인 API 작성
  - http://3.36.66.115/best
  - http://3.36.66.115/main
  - http://3.36.66.115/course
  - http://3.36.66.115/soup
  - http://3.36.66.115/side
  - http://3.36.66.115/detail/{hash}
- NginX에 배포 - http://3.36.66.115/
  - 톰캣 프록시 설정
  - 정적 페이지는 NginX에서 관리하도록 설정
  - <details>
        <summary>서버 설정 코드</summary>
        
        server {
            listen 80 default_server;
            listen [::]:80 default_server;

            root /var/www/html;

            server_name localhost;
    
            location / {
                    try_files $uri $uri/ @apachesite;
            }
    
            location ~ \.(css|js|jpg|jpeg|gif|htm|html|swf)$ {
            }
    
            location @apachesite {
                    proxy_set_header X-Real-IP  $remote_addr;
                    proxy_set_header Host $host;
                    proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
                    proxy_pass http://127.0.0.1:8081;
            }
            
            # try_files 테스트용
            location /index2 {
                    try_files $uri /index.html;
            }
        }
        
    </details>

- 개발환경과 배포환경 분리
  - dev와 prod로 프로필 나눠서 설정 서로 다르게 관리
  - 현재 dev만 사용

gradle 태스크를 작성하다 잘 안되는 부분이 있었습니다.
아래와 같이 스크립트를 작성하니 빌드시에 해당 스크립트까지 같이 실행이 되는 문제가 발생했습니다.
찾아봐도 해결점을 못찾겠어서 현재 해당 내용은 없앤 상태입니다.
```groovy
task bootRun_prod(type: Task) {
    tasks.bootRun.exec()
}
```
이외에 직접 커맨드라인으로 실행하는 방법도 생각해봤는데, 여러 환경을 함께 지원하거나 모듈로 프로젝트를 열었을 경우 처리하기가 힘들어 포기했습니다.

감사합니다.

Dae-Hwa committed:
[DOCS] : README.md에 그라운드 룰, 커밋 룰, 요구사항 분석 추가

노션에 정리

Dae-Hwa committed:
[CHORE] : .gitignore 추가

Dae-Hwa committed:
[CHORE] : 프로젝트 초기화

Dae-Hwa committed:
[CHORE] : .gitignore에 .logs 추가

BE 로그파일 저장 경로 설정

Dae-Hwa committed:
[CHORE] : 로그 세팅

logs/sidedish/.. 에 log 저장

Dae-Hwa committed:
[CHORE] : H2 설정

H2설정 기본으로 잡은 뒤, 운영용 설정 따로 잡을 예정

Dae-Hwa committed:
[FIX] : SidedishBeApplicationTests 패키지 변경

example로 되어 있어 기본 패키지(com.codesquad.xx)와 일치하지 않는 문제
수정

Dae-Hwa committed:
[CHORE] : dev와 prod로 설정 분리

application.properties 에서 common.properties 참조하도록 설정. 추가적인
설정은 필요시 추가

Dae-Hwa committed:
[CHORE] : prod에 mysql 세팅 추가

localhost 연결 확인 완료

Dae-Hwa committed:
[CHORE] : bootRun_prod 추가

Dae-Hwa committed:
[CHORE] : 포트 8081로 변경

Dae-Hwa committed:
[CHORE] : bootrun_prod 삭제 및 build_prod 수정

bootrun_prod 사용 시 build_prod 실행 불가. build_prod 실행 시 gradlew
경로를 제대로 찾지 못하는 문제 수정

Dae-Hwa committed:
[FEAT] : 테스트용 hello 컨트롤러 추가

/hello 로 접속 시 "hello" 리턴

Jiwon-JJW committed:
Merge pull request #4 from Dae-Hwa/BE/preparation/project-setup

[Be]project setup

Dae-Hwa committed:
[CHORE] : jackson 프로퍼티 네이밍 정책 Snake Case로 설정

api문서에 맞게 수정

Dae-Hwa committed:
[FEAT] : SidedishDTO 생성 및 json 읽어서 객체 생성할 수 있는 기능 추가

jackson object mapper 사용. json은 String 혹은 파일로 입력 가능. json파일은 개발 환경에만 들어갈 것이므로 Environment를 이용하여 환경에 따라 테스트 로직 실행

Dae-Hwa committed:
[FEAT] : SampleDataFactory 추가

샘플 데이터를 json파일로 읽어서 가져온다. ClassPathResource 이용하여 resource 폴더 내의 json파일 읽어옴. 추후 리팩토링 가능해보임. dev일때만 등록되게 해보는 방안도 고려해볼 수 있을 것 같다.

Dae-Hwa committed:
[FEAT] : BestSidedishController 추가

베스트반찬 출력해주는 목업 api 추가. 현재는 SampleDataFactory에서 가져오도록 작성.

Dae-Hwa committed:
[REFACTOR] : 불필요한 import 제거

Dae-Hwa committed:
[REFACTOR] : 불필요한 중복 제거

Jiwon-JJW committed:
Merge pull request #7 from Dae-Hwa/BE/preparation/mockup-api

[Be]setup mockup api

Jiwon-JJW committed:
[ADD] : upload sample Json file

Mockup API 내의 sample data를 Json으로 만들었다.

Jiwon-JJW committed:
[ADD] : Json 파일 읽어내는 기능 추가

Json 파일의 내용을 읽어내어 확인하는 기능까지 추가하였다.

Jiwon-JJW committed:
[FIX] : 더 좋은 변수명으로 변경

Jiwon-JJW committed:
[ADD] : Json파일을 읽어내어 Main 반찬을 생성하는 기능 추가

ClassPathResource를 이용하여 Json파일을 읽어, Main 반찬을 List로 생성한다.

Jiwon-JJW committed:
[REFACTOR] : DTO에 있던 Factory메서드의 중복을 제거

ObjectMapper로 Json 변환하는 메서드가 중복되어 하나로 합쳤다. - 제네릭 타입 메서드
메서드의 중복이 제거되어, Test Class들도 하나로 합쳐 테스트의 중복도 제거했다.

Jiwon-JJW committed:
[ADD] : SampleDataFactory class에 private 생성자 생성.

다른 클래스에서 static 메서드만 사용하므로, private 생성자를 만들어 주었다.

Jiwon-JJW committed:
[FIX] : 불필요한 줄바꿈 제거

SampleDataFactory와 SampleDataFactoryTest에 있던 불필요한 줄바꿈들을 제거해주었다.

Jiwon-JJW committed:
[FIX] : 파일 끝에 개행문자 추가.

main-sidedishes.json의 파일 끝에 개행문자가 빠져있어, 추가해주었다.

Dae-Hwa committed:
Merge pull request #9 from Jiwon-JJW/BE/preparation/mockup-api

Main 반찬 파일 변환 및 메서드 리팩토링 작업

Jiwon-JJW committed:
[ADD] : Main 반찬에 대한 Controller 생성

주소에 "/main" 입력 시, Main 반찬 목록을 보여주는 기능을 추가한 Controller를 생성했다.

Jiwon-JJW committed:
[FIX] : 간격이 맞지 않는 것에 대해 정렬 작업

ItemDTOTestResults의 MAIN_MENU 내의 간격이 맞지않아 정렬 작업 진행.

Jiwon-JJW committed:
[ADD] : Course 반찬의 Json 파일 확인하는 기능 추가

Course 반찬의 Json파일을 확인하는 테스트 기능을 SampleDataFactoryTest에 추가하였다.

Jiwon-JJW committed:
[ADD] : Json파일을 읽어내어 Course 반찬을 생성하는 기능 추가

Json파일이 있는 경로를 입력하여 Json파일을 찾아내, 그 것을 기반으로 Course 반찬을 List로 생성한다.

Jiwon-JJW committed:
[ADD] : Course 반찬에 대한 Controller 생성

주소에 "/course" 입력 시, Course 반찬 목록을 보여주는 기능을 추가한 Controller를 생성했다.

Jiwon-JJW committed:
[FIX] : 파일 끝에 개행문자 추가.

json 파일 끝에 개행문자가 빠져있어, 추가해주었다.

Jiwon-JJW committed:
[ADD] : Soup 반찬의 Json 파일 확인하는 기능 추가

Soup 반찬의 Json파일을 확인하는 테스트 기능을 SampleDataFactoryTest에 추가하였다.

Jiwon-JJW committed:
[ADD] : Json파일을 읽어내어 Soup 반찬을 생성하는 기능 추가

Json파일이 있는 경로를 확인하여 Json파일을 찾아내, 그 것을 기반으로 Soup 반찬을 List로 생성한다.

Dae-Hwa committed:
[ADD] : Detail을 모아둔 json파일 및 DTO 생성하여 파싱

detail은 하나씩 뽑아야 하므로, map에서 해당 해쉬 키로 뽑아 쓸 수 있도록 구성. 추후 repository에 넣을 시 수정 필요할 수 있음.

Jiwon-JJW committed:
[ADD] : Soup 반찬에 대한 Controller 생성

주소에 "/soup" 입력 시, Soup 반찬 목록을 보여주는 기능을 추가한 Controller를 생성했다.

Jiwon-JJW committed:
[ADD] : Side 반찬의 Json 파일 확인하는 기능 추가

Side 반찬의 Json파일을 확인하는 테스트 기능을 SampleDataFactoryTest에 추가하였다.

Jiwon-JJW committed:
[ADD] : Json파일을 읽어내어 Side 반찬을 생성하는 기능 추가

Json파일이 있는 경로를 확인하여 Json파일을 찾아내, 그 것을 기반으로 Side 반찬을 List로 생성한다.

Jiwon-JJW committed:
[ADD] : Side 반찬에 대한 Controller 생성

주소에 "/side" 입력 시, Side 반찬 목록을 보여주는 기능을 추가한 Controller를 생성했다.

Jiwon-JJW committed:
[REFACTOR] : DTO 클래스와 Controller 클래스의 패키지 분리.

Controller의 개수가 늘어났으므로, DTO클래스와 Controller 클래스를 각각의 패키지로 분리해주었다.

Dae-Hwa committed:
[ADD] : Price 추가 및 반영

금액 관련 getter 와 setter에서 반복적으로 NumberFormat 사용되어 중복 제거. getter에서는 price.getFormattedPrice 사용하도록 작성. 기본 postfix "원"으로 설정

Dae-Hwa committed:
[TEST] : SampleDataFacotry의 createDtails테스트 코드 추가

Dae-Hwa committed:
[ADD] : DeatilController 추가

추후 컨트롤러 모이면 통합해보는 것 고려

Dae-Hwa committed:
[FIX] : JsonInclude 제거. 추후 논의할 사항

Dae-Hwa committed:
Merge pull request #10 from Jiwon-JJW/BE/preparation/mockup-api-main

[BE] Main 반찬 컨트롤러 추가

Dae-Hwa committed:
Merge pull request #11 from Jiwon-JJW/BE/preparation/mockup-api-course

[BE] Course 반찬 추가

Dae-Hwa committed:
Revert "[BE] Main 반찬 컨트롤러 추가"

Dae-Hwa committed:
Revert "[BE] Course 반찬 추가"

Dae-Hwa committed:
Merge pull request #15 from Dae-Hwa/revert-10-BE/preparation/mockup-api-main

Revert "[BE] Main 반찬 컨트롤러 추가"

Dae-Hwa committed:
Merge pull request #16 from Dae-Hwa/revert-11-BE/preparation/mockup-api-course

Revert "[BE] Course 반찬 추가"

Jiwon-JJW committed:
Merge branch 'BE/preparation/mockup-api' into BE/preparation/mockup-api-side

Dae-Hwa committed:
Merge pull request #13 from Jiwon-JJW/BE/preparation/mockup-api-side

[BE] Side 반찬 추가 및 패키지 분리

Dae-Hwa committed:
Merge branch 'BE/preparation/mockup-api' into BE/preparation/mockup-api-detail

# Conflicts:
#	be/sidedish-be/src/main/java/com/codesquad/sidedish/utils/SampleDataFactory.java
#	be/sidedish-be/src/test/java/com/codesquad/sidedish/utils/SampleDataFactoryTest.java

Dae-Hwa committed:
Merge branch 'BE/preparation/mockup-api' into BE/preparation/mockup-api-detail

# Conflicts:
#	be/sidedish-be/src/main/java/com/codesquad/sidedish/utils/SampleDataFactory.java
#	be/sidedish-be/src/test/java/com/codesquad/sidedish/utils/SampleDataFactoryTest.java

Jiwon-JJW committed:
Merge pull request #14 from Dae-Hwa/BE/preparation/mockup-api-detail

[Be] mockup api-detail

Dae-Hwa committed:
Merge pull request #17 from Dae-Hwa/BE/preparation/mockup-api

[Be] preparation/mockup api

Dae-Hwa committed:
[FIX] : 배포 시 JSON파일 읽을 수 없는 문제 수정

ClassPathResource 사용 시 URL이 File프로토콜을 사용하는지 확인한다.
하지만 executable jar로 실행 시, URL이 jar:file:로 되어 file프로토콜로
인식하지 못한다. 따라서 InputStream으로 수정.

Closes #18

Dae-Hwa committed:
[FIX] : json파일 변환 에러 메세지 더 상세하게 작성

에러 유추가 힘들어 새롭게 작성.

Dae-Hwa committed:
Merge pull request #19 from Dae-Hwa/BE/error-fix/cannot-find-jsonfile

[FIX] : 배포 시 JSON파일 읽을 수 없는 문제 수정

ksundong commented on be/sidedish-be/build.gradle:
```diff
@@ -0,0 +1,54 @@
+}
+
+group = 'com.codesquad'
+version = ''
```
버전이 없네요??

Dae-Hwa replied:
제가 이해도가 떨어져서 그런데, 메인 브랜치에 머지되거나 하는 등의 이유로 버전업이 됐다고 가정했을때, 그 때마다 build.gradle도 수정하는 식으로 관리해줘야하는건가요?

ksundong replied:
그건 정하기 나름이죠~
정답은 없습니다.

ksundong commented on be/sidedish-be/build.gradle:
```diff
@@ -0,0 +1,54 @@
+    useJUnitPlatform()
+}
+
+ext.profile = (!project.hasProperty('profile') || !profile) ? 'dev' : profile
+
+sourceSets {
+    main {
+        java {
+        }
+        resources {
+            srcDirs += ["src/main/resources-${profile}"]
+        }
+    }
+}
+
+task build_prod(type: GradleBuild) {
+    startParameter.setTaskNames(["build"])
+    startParameter.setProjectProperties(new HashMap() {
+        {
+            put("profile", "prod")
+        }
+    })
+}
```
이렇게 하지않고도, 외부에서 충분히 profile을 select 해줄 수 있을텐데요..🤔

Dae-Hwa replied:
환경별로 아예 파일 관리를 다르게 하고 싶었습니다. 특히 json파일과 같이 개발환경에서만 사용되어야 하는 파일들을 따로 관리를 하고 싶었습니다. 
결론적으로는 빌드하여 서버에 올릴 jar파일 내부에 꼭 필요한 파일과 설정만 배포하고 싶어서 이런 구성을 했습니다.
Spring 내부에서 프로필을 선택하거나 jar 실행 시 프로필을 선택하도록 하지 않은 것은 위와 같이 구성할 수 없다고 판단하여 그렇게 했습니다.

ksundong replied:
저는 한 번도 안해봤던 고민이네요. 어떤 의도로 작성하신건지 알 것 같아요.
- 개발환경에서만 사용되는 파일은 gitignore에 등록하면 되지 않나요? 빌드는 정석대로라면 빌드서버를 따로 두고 빌드를 하는게 맞으니까요.

Dae-Hwa replied:
말씀해주신 부분이 맞긴한데, 배포 시에 숨기고 싶은 자원(json파일이나 sql파일 등)도 공유가 됐으면 해서 이런 선택을 했습니다. 
아래에 답변 주신 부분과 함께 생각해보겠습니다!

ksundong commented on be/sidedish-be/src/main/java/com/codesquad/sidedish/temp/HelloController.java:
```diff
@@ -0,0 +1,12 @@
+import org.springframework.web.bind.annotation.GetMapping;
+import org.springframework.web.bind.annotation.RestController;
+
+@RestController
+public class HelloController {
+    @GetMapping("/hello")
+    public String hello() {
+        return "hello";
+    }
+}
```
안쓰면 지워주세요~

Jiwon-JJW replied:
지워주었습니다! 567dac1

ksundong commented on be/sidedish-be/src/main/java/com/codesquad/sidedish/utils/SampleDataFactory.java:
```diff
@@ -0,0 +1,68 @@
+
+public class SampleDataFactory {
+
+    private SampleDataFactory() {
+
+    }
```
```suggestion
    private SampleDataFactory() {
    }
```

빈 body는 비워주는 것이 흔히 사용하는 컨벤션입니다.

ksundong commented on be/sidedish-be/src/main/java/com/codesquad/sidedish/utils/SampleDataFactory.java:
```diff
@@ -0,0 +1,68 @@
+
+        try (InputStream is = new ClassPathResource(jsonFilePath).getInputStream()) {
+            return objectMapper.readValue(is, typeReference);
+        } catch (JsonParseException e) {
+            throw new IllegalStateException("Json파일 파싱 중 에러 발생. jsonFilePath: " + jsonFilePath, e);
+        } catch (IOException e) {
+            throw new IllegalStateException("파일 읽는 중 에러 발생. jsonFilePath: " + jsonFilePath, e);
+        }
```
예외 포장 좋은데요, Custom예외의 사용을 아끼진 않았으면 좋겠네요!

Jiwon-JJW replied:
Custom예외를 두 개로 만들어 처리 진행 했습니다!
c7b3174
다만 각각 다른 상황에서 발생하는 예외라고 생각되어 두 개로 만들었는데, 하나로 하는게 더 나은 방향인지 궁금합니다..!

ksundong replied:
다른 커밋이네요 ㅋㅋ

ksundong replied:
예외는 여러개 만들어도 상관없고, 둘이 처리가 같다면 같은 부모 예외를 두면 되겠죠?

ksundong commented on be/sidedish-be/src/main/java/com/codesquad/sidedish/utils/SampleDataFactory.java:
```diff
@@ -0,0 +1,68 @@
+        ObjectMapper objectMapper = new ObjectMapper();
+        objectMapper.setPropertyNamingStrategy(PropertyNamingStrategy.SNAKE_CASE);
+
+        try (InputStream is = new ClassPathResource(jsonFilePath).getInputStream()) {
```
try-with-resources의 사용 좋습니다.
is는 별로 좋지 않은 네이밍이지만, 흔히 사용되니 넘어가죠~

ksundong commented on be/sidedish-be/src/main/java/com/codesquad/sidedish/utils/SampleDataFactory.java:
```diff
@@ -0,0 +1,68 @@
+    }
+
+    private static <E> E jsonToObject(String jsonFilePath, TypeReference<E> typeReference) {
+        ObjectMapper objectMapper = new ObjectMapper();
+        objectMapper.setPropertyNamingStrategy(PropertyNamingStrategy.SNAKE_CASE);
```
이 ObjectMapper는 호출 될 때마다 새로 생성될텐데, 이를 해결할 방법은 어떤 것이 있을까요?

Jiwon-JJW replied:
ObjectMapper를 static 변수로 새로 생성하여 이를 해결해 보았습니다!
c7b3174

ksundong replied:
스프링에서 제공하는 Bean Factory를 활용해주시는건 어떨까요?

ksundong commented on be/sidedish-be/src/main/java/com/codesquad/sidedish/utils/SampleDataFactory.java:
```diff
@@ -0,0 +1,68 @@
+
+    }
+
+    public static List<SidedishDTO> createBestSidedishes() {
+        return jsonToObject("sample-data/best-sidedishes.json", new TypeReference<List<SidedishDTO>>() {
+        });
+    }
+
+    public static List<ItemDTO> createMainSidedishes() {
+        return jsonToObject("sample-data/main-sidedishes.json", new TypeReference<List<ItemDTO>>() {
+        });
+    }
+
+    public static List<ItemDTO> createCourseSidedishes() {
+        return jsonToObject("sample-data/course-sidedishes.json", new TypeReference<List<ItemDTO>>() {
+        });
+    }
+
+    public static List<ItemDTO> createSoupSidedishes() {
+        return jsonToObject("sample-data/soup-sidedishes.json", new TypeReference<List<ItemDTO>>() {
+        });
+    }
+
+    public static List<ItemDTO> createSideSidedishes() {
+        return jsonToObject("sample-data/side-sidedishes.json", new TypeReference<List<ItemDTO>>() {
+        });
+    }
```
비슷한 형태가 반복됩니다. 중복 제거를 해볼 수 있겠네요.

Jiwon-JJW replied:
말씀대로 중복제거를 최대한 진행해보았습니다!
26ff085

ksundong commented on be/sidedish-be/src/main/java/com/codesquad/sidedish/web/sidedish/DTO/DetailDTO.java:
```diff
@@ -0,0 +1,216 @@
+        private DetailDTOBuilder() {
+        }
+
+        public static DetailDTOBuilder create() {
+            return new DetailDTOBuilder();
+        }
```
이 메서드를 `DetailDTO` 에 둔다면 이런 스타일이 될겁니다.

```java
public static DetailDTOBuilder builder() {
    return new DetailDTOBuilder();
}
```

구현해주신 형태도 좋습니다만, 위의 형태도 고려해봐주세요.

ksundong replied:
`DetailDTO.builder()...` 과 같은 형태로 사용할 수 있습니다.

Dae-Hwa replied:
추가해줬습니다! 3f7d131

ksundong replied:
create를 굳이 사용하지 않아도 되겠쬬?

ksundong commented on be/sidedish-be/src/main/java/com/codesquad/sidedish/web/sidedish/DTO/DetailDTO.java:
```diff
@@ -0,0 +1,216 @@
+        }
+    }
+
+    public static class Data {
```
내부 클래스 사용보다는, 외부로 빼주는 것이 좋을 것 같은데요~

Dae-Hwa replied:
분리했습니다!
8a8bdca

ksundong replied:
이름까지 잘 바꿔주셨지만, 일반적인 이름의 사용은 피하는게 좋습니다.

ksundong commented on be/sidedish-be/src/main/java/com/codesquad/sidedish/web/sidedish/DTO/DetailDTO.java:
```diff
@@ -0,0 +1,216 @@
+
+    public static class Data {
+        private String topImage;
+        private List<String> thumbImages;
+        private String productDescription;
+        private Price point;
+        private String deliveryInfo;
+        private Price deliveryFee;
+        private List<Price> prices;
+        private List<String> detailSection;
```
필드가 어떤 순서로 배치된건지 이해하기 어렵네요.
그리고, Collection들은 `null`로 초기화하기 보다는 Empty Collection으로 두는 것이 좋다고 생각합니다.

Dae-Hwa replied:
empty list로 초기화해줬습니다
530f271

ksundong commented on be/sidedish-be/src/main/java/com/codesquad/sidedish/web/sidedish/DTO/ItemDTO.java:
```diff
@@ -0,0 +1,172 @@
+public class ItemDTO {
+    private String detailHash;
+    private String image;
+    private String alt;
```
alt는 어떤 데이터가 담기나요? 별로 구체적이지 않은 것 같아요.
이렇게 협의가 된건가요?

ksundong commented on be/sidedish-be/src/main/java/com/codesquad/sidedish/web/sidedish/DTO/ItemDTO.java:
```diff
@@ -0,0 +1,172 @@
+    private List<String> deliveryType;
+    private String title;
+    private String description;
+    private Price nPrice;
+    private Price sPrice;
```
n과 s도 마찬가지입니다. 별로 좋은 네이밍은 아닌 것 같네요.

ksundong commented on be/sidedish-be/src/main/java/com/codesquad/sidedish/web/sidedish/DTO/ItemDTO.java:
```diff
@@ -0,0 +1,172 @@
+        this.description = description;
+    }
+
+    public String getnPrice() {
+        return nPrice != null ? nPrice.getFormattedPrice() : "";
+    }
```
저는 이런 형태의 구현은 별로 선호하지 않아요.
getter는 getter의 역할을 해야하지, 별도의 조작이 가해진다면, getter가 아니라고 생각합니다.

Dae-Hwa replied:
지금 상황에서 이렇게 구현하지 않으면, 생성자와 setter에 모두 해당 소스가 들어가야돼서 이런 식으로 구현했던 것 같습니다.
엔티티 객체들이 구현되면 자연스럽게 해결될 것 같은데, 말씀해주신 부분 꼭 생각하며 진행해보겠습니다 감사합니다!

ksundong replied:
```suggestion
    @JsonProperty("n_price")
    public String formattednPrice() {
        return nPrice != null ? nPrice.getFormattedPrice() : "";
    }
```

ksundong replied:
이런 식의 구현은 어떤가요?

Dae-Hwa replied:
아... 이런 방법은 생각을 아예 못 해봤네요.. 훨씬 괜찮아 보입니다! 적용해봐야겠어요 ㅎㅎ 감사합니다

ksundong commented on be/sidedish-be/src/main/java/com/codesquad/sidedish/web/sidedish/Price.java:
```diff
@@ -0,0 +1,33 @@
+import java.text.NumberFormat;
+import java.text.ParseException;
+
+public class Price {
```
클래스의 분리 좋습니다.
근데 이건 값 객체로 쓸 용도였나요? `price` 필드를 `final`로 정의해보는건 어떨까요?

ksundong replied:
그리고 단위 테스트도 작성해보기 적절한 클래스 같아보이는데 한 번 도전해보시는 것을 추천합니다.

Dae-Hwa replied:
final로 변경해줬습니다! 436c959

ksundong commented on be/sidedish-be/src/main/java/com/codesquad/sidedish/web/sidedish/controller/BestSidedishController.java:
```diff
@@ -0,0 +1,16 @@
+public class BestSidedishController {
+    @GetMapping("/best")
+    public List<SidedishDTO> readBestSidedishes() {
+        return SampleDataFactory.createBestSidedishes();
```
매 요청마다 이걸 만드는 형태가 될 것 같은데요?

ksundong commented on be/sidedish-be/src/main/resources-dev/data.sql:
```diff
@@ -0,0 +1 @@
+INSERT INTO test_table(TEST_VALUE) VALUES ('test')
```
필요한가요?

ksundong commented on be/sidedish-be/src/main/resources-prod/application.properties:
```diff
@@ -0,0 +1,13 @@
+spring.profiles.active=prod
+spring.config.import=classpath:common.properties
+
+# logging
+logging.level.com.codesquad.esfj.todolist.*=DEBUG
+logging.level.root=info
+
+# mysql
+spring.datasource.driver-class-name=com.mysql.cj.jdbc.Driver
+spring.datasource.url=jdbc:mysql://localhost:3306/sidedish?serverTimezone=UTC&characterEncoding=UTF-8
+spring.datasource.username=sidedish
+spring.datasource.password=sidedish
+spring.datasource.initialization-mode=always
```
이렇게 application properties를 만드는 것이 아닙니다.
https://docs.spring.io/spring-boot/docs/current/reference/html/spring-boot-features.html#boot-features-external-config
여기 참고해서 만들어보세요.

Dae-Hwa replied:
https://github.com/codesquad-members-2021/sidedish/pull/22#discussion_r619984231
위 내용에 이어지는 내용같아 커맨트 링크를 걸었습니다. 저런 이유로 gradle과 resource 디렉토리 및 application.properties를 이와 같이 구성했는데요.

gradle 스크립트로 빌드 시 패키징할 자원을 선택하는건 좋지 않은 방법인건가요?

application properties를 이렇게 만들면 안 된다고 하셨는데, 이해도가 떨어지다보니 공식 문서만을 봐서는 왜 그런지 유추하기가 힘든 것 같습니다.

Dae-Hwa replied:
https://github.com/codesquad-members-2021/sidedish/pull/33#discussion_r619667076 이렇게 코맨트 남겨 주신걸 봐서 다시 생각해봤습니다.
war배포가 아니니 jndi는 의미가 없을 것 같고, 암호화도 결국 소스에 비밀번호가 포함되니 옳은 방법은 아닌 것 같습니다.
그렇다면 서버에 미리 계정정보가 담긴 파일같은걸 넣어두고 그걸 읽어올 수 있도록 경로설정을 해주는 방식으로 구현하는게 현재로서 최선일 것 같은데 혹시 이런 접근이 맞는것인가요?

ksundong replied:
> [#22 (comment)](https://github.com/codesquad-members-2021/sidedish/pull/22#discussion_r619984231)
> 위 내용에 이어지는 내용같아 커맨트 링크를 걸었습니다. 저런 이유로 gradle과 resource 디렉토리 및 application.properties를 이와 같이 구성했는데요.
> 
> gradle 스크립트로 빌드 시 패키징할 자원을 선택하는건 좋지 않은 방법인건가요?
> 
> application properties를 이렇게 만들면 안 된다고 하셨는데, 이해도가 떨어지다보니 공식 문서만을 봐서는 왜 그런지 유추하기가 힘든 것 같습니다.

일단, 너무 지엽적인 부분이라고 생각이 됩니다. 배포 물론 중요한데, Application 개발이 먼저라고 생각되네요.

제 생각에는, build script로 제외할 만큼의 리소스가 없다고 생각됩니다.
좋다, 좋지 않다의 판단이라기 보다는, '그럴 필요가 없다.' 정도로 생각해주시면 되겠습니다.
패키징 된 파일을 배포할게 아닌이상에야 json 파일이 있고 없고는 큰 의미가 없다고 생각합니다.

공식문서를 보여드린 이유는, Spring에서도 패키지 분리를 염두에 두고 있지 않은 것 같기 때문입니다. 만약 개별 packaging이 필요한 경우 직접 구현해라~ 정도일 것이고, 일반적으로는 한 패키지를 실행할 때, 환경변수를 지정해 실행하는 것이 일반적입니다.
application.properties는 경로를 분리하지 않고 사용하는 것이 일반적이기에, 필요하다면 필요한 리소스만 별도로 패키징 로직을 타는게 맞다고 판단됩니다.

ksundong replied:
> [#33 (comment)](https://github.com/codesquad-members-2021/sidedish/pull/33#discussion_r619667076) 이렇게 코맨트 남겨 주신걸 봐서 다시 생각해봤습니다.
> war배포가 아니니 jndi는 의미가 없을 것 같고, 암호화도 결국 소스에 비밀번호가 포함되니 옳은 방법은 아닌 것 같습니다.
> 그렇다면 서버에 미리 계정정보가 담긴 파일같은걸 넣어두고 그걸 읽어올 수 있도록 경로설정을 해주는 방식으로 구현하는게 현재로서 최선일 것 같은데 혹시 이런 접근이 맞는것인가요?

이 역시도 맞다. 맞지 않다로 분리될 수 있는 것이 아닙니다. ㅎㅎ DB 접속정보를 제공하는 많은 방법이 있겠죠.
중요한 것은 하면 안되는 일을 하지 않는 것이라고 생각합니다.

즉, 말씀하신 대로 비밀번호는 노출되지 않아야 하고, 이를 지키는 방향으로 구현해주시면 됩니다.

Dae-Hwa replied:
정답이 없는 문제라 많이 어려운 것 이었네요 ㅎㅎ.. 
저희 둘 다 엔진엑스를 처음 써보다 보니 서버를 띄워보는 것을 목표로 했더니 너무 지엽적인 부분들에 신경을 많이 써버린 것 같기도 하네요
말씀해주신 부분들 참고해서 다시 한 번 생각해보겠습니다! 자세히 알려주셔서 감사합니다!!

Dae-Hwa replied:
디렉토리 구조를 기본설정으로 돌려줬습니다.
resources 안에 application.properties를 넣어주고, application-default, application-prod 와 같이 프로필에 따라 프로퍼티 파일을 선택할 수 있도록 해줬습니다.
DB접속정보는 소스에 담아두지 않고, 프로젝트 루트(java 실행 위치)에 있는 dbcofig.properties로 읽어오도록 했습니다. 따라서 깃에 함께 담기지 않고, 서버에 미리 가져다 놓고 사용할 예정입니다.

ksundong replied:
서버에 properties를 두어도 좋고, 환경변수로 두어도 됩니다.
방법은 여러가지에요 ㅎㅎ

ksundong commented on be/sidedish-be/src/test/java/com/codesquad/sidedish/utils/SampleDataFactoryTest.java:
```diff
@@ -0,0 +1,110 @@
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+@SpringBootTest
```
이 애노테이션은 어떨 때 사용하는 걸까요?

Dae-Hwa replied:
스프링의 컨텍스트가 필요할 때 사용하는 것인데, 필요없어졌기 때문에 제거해줬습니다

ksundong commented on be/sidedish-be/src/test/java/com/codesquad/sidedish/utils/SampleDataFactoryTest.java:
```diff
@@ -0,0 +1,110 @@
+@SpringBootTest
+class SampleDataFactoryTest {
+
+    @Autowired
+    Environment environment;
```
과연 이게 필요했는지는 의문이네요.

Dae-Hwa replied:
제거했습니다! 4eea66b33d4bfb0f1a43be7e98f951804eeea5c0

ksundong commented on be/sidedish-be/src/test/java/com/codesquad/sidedish/utils/SampleDataFactoryTest.java:
```diff
@@ -0,0 +1,110 @@
+
+    @Test
+    void createBestSidedishes() throws JsonProcessingException {
+        if (!environment.acceptsProfiles(Profiles.of("dev"))) {
+            return;
+        }
```
이 로직도 필요없죠 사실?

Dae-Hwa replied:
제거했습니다!

ksundong commented on be/sidedish-be/src/test/java/com/codesquad/sidedish/utils/SampleDataFactoryTest.java:
```diff
@@ -0,0 +1,110 @@
+
+    static Stream<Arguments> createDetails() {
+        return Stream.of(
+                Arguments.of("H9881", DetailDTOTestResults.H9881)
```
Case를 좀 더 넣어줍시다!

Jiwon-JJW replied:
케이스 입력 진행했습니다! 감사합니다!
283cb52

ksundong commented on be/sidedish-be/src/test/java/com/codesquad/sidedish/utils/SampleDataFactoryTest.java:
```diff
@@ -0,0 +1,110 @@
+    }
+
+
+    @ParameterizedTest
+    @MethodSource
```
👍 
다만, MethodSource로 사용하는 메서드는 이름을 바꿔주는게 더 좋을 것 같네요.

Dae-Hwa replied:
be87641 변경해줬습니다!

ksundong commented on be/sidedish-be/src/test/java/com/codesquad/sidedish/web/sidedish/DetailDTOTestResults.java:
```diff
@@ -0,0 +1,17 @@
+package com.codesquad.sidedish.web.sidedish;
+
+public class DetailDTOTestResults {
+    public static final String H9881 = "{" + System.lineSeparator() +
+            "  \"hash\" : \"H9881\"," + System.lineSeparator() +
+            "  \"data\" : {" + System.lineSeparator() +
+            "    \"top_image\" : \"https://cdn.bmf.kr/_data/product/H9881/910a01a81c49cb75414edb759237501f.jpg\"," + System.lineSeparator() +
+            "    \"thumb_images\" : [ \"https://cdn.bmf.kr/_data/product/H9881/910a01a81c49cb75414edb759237501f.jpg\", \"https://cdn.bmf.kr/_data/product/H9881/fbf29077698ca16f8050e43476b47f38.jpg\", \"https://cdn.bmf.kr/_data/product/H9881/c96c6949efc3391148e9b280a2c5ed0b.jpg\", \"https://cdn.bmf.kr/_data/product/H9881/71411e15d2d961df496f87f08648b345.jpg\", \"https://cdn.bmf.kr/_data/product/H9 ...
+            "    \"product_description\" : \"경상도 명물 요리 세 가지를 한 상에!\"," + System.lineSeparator() +
+            "    \"point\" : \"312원\"," + System.lineSeparator() +
+            "    \"delivery_info\" : \"서울 경기 새벽배송 / 전국택배 (제주 및 도서산간 불가) [화 · 수 · 목 · 금 · 토] 수령 가능한 상품입니다.\"," + System.lineSeparator() +
+            "    \"delivery_fee\" : \"2,500원 (40,000원 이상 구매 시 무료)\"," + System.lineSeparator() +
+            "    \"prices\" : [ \"39,000원\", \"31,200원\" ]," + System.lineSeparator() +
+            "    \"detail_section\" : [ \"https://cdn.bmf.kr/_data/product/H9881/7fb1ddf1adeadc5410cecd79441f7b65.jpg\", \"https://cdn.bmf.kr/_data/product/H9881/b776c59544b516a184d1363c2c802789.jpg\", \"https://cdn.bmf.kr/_data/product/H9881/cc2b4a61db410096db0e3c497096d63f.jpg\", \"https://cdn.bmf.kr/_data/product/H9881/77970960c8efe0992f9746c37062e1e4.jpg\", \"https://cdn.bmf.kr/_data/product/ ...
+            "  }" + System.lineSeparator() +
+            "}";
```
과연 이걸 다 확인하는게 맞을지는 잘 모르겠네요.
그냥 객체 하나 만들어서 ObjectMapper로 만드는게 나았을지도요..

ksundong reviewed (changes requested):
안녕하세요! 프레디 & Starve!! 리뷰어 Dion입니다.
Mock API를 정성스럽게 작성해주셨는데, 아쉬운 부분이 많네요.
저도 작년에 비슷한 프로젝트를 진행했던 만큼 제가 프로젝트를 진행했던 경험을 바탕으로 들었던 아쉬운 점을 적어볼게요.

- 데이터를 왜 그대로 가져와서 그대로 사용하는 형태로 정의해주셨나요? 좋지 않은 네이밍이 존재하는 형태인데, 그대로 사용해주신 부분이 의아했습니다.
- Mock API 작성에만 굉장히 많은 시간이 들어간 것 같은데요, 저라면 간단하게 구현 후, 실제 API를 빠르게 구현시작하는 형태로 진행했을 것 같습니다. 실제로도 그렇게 했고요.
- 사실 데이터만 전달하는게 끝이 아니라, API 설계가 들어갔으면 했는데, 아쉽네요.

이건 제가 단순히 아쉬웠던 부분이니, 그대로 진행해주시면 됩니다.

아래의 코멘트 참고해주시고, 궁금한 점이나, 도움이 필요한 부분, 이해가 되지 않는 부분, 토론하고 싶은 부분이 있으면 언제든 코멘트 남겨주세요. 고생하셨습니다~

Dae-Hwa committed:
[REFACTOR] : Price의 필드 final로 변경

값 객체이기때문에 final로 해도 동작해야한다.

Dae-Hwa committed:
[REFACTOR] : Data 내부객체에서 별도의 DTO로 분리

Dae-Hwa committed:
[Refactor] : DTO내부의 List emptyList로 초기화

Jiwon-JJW committed:
[FIX] : Json 파일 경로와 ObjectMapper 대한 리팩토링 진행.

Json 파일 경로 중, 중복되는 두 종류의 글자를 상수로 분리하였고, ItemDTO를 담는 List 또한 중복을 제거하기 위하여 메서드를 생성하였다.
ObjectMapper는 매번 생성하지 않도록, 따로 분리하여 객체를 생성해주었다.

Jiwon-JJW committed:
[ADD] : Custom Exception 생성.

Json 파싱과 파일을 읽어올 수없을때 발생하는 예외를 처리하기 위한 Custom Exception을 생성하였다.

Jiwon-JJW committed:
[DELETE]  : 사용하지 않는 test 클래스와 sql문 삭제

테스트를 위한 HelloController와 data 삭제

Dae-Hwa committed:
[Refactor] : builder 메소드 추가

Dae-Hwa committed:
[REFACTOR] : Methodsource 이름 변경

Dae-Hwa committed:
[FIX] : 일치하지 않는 테스트 케이스 수정

Dae-Hwa committed:
[CHORE] : build.gradle 및 resources 설정 변경.

profile jar실행시 명시해주도록 변경. resources 구조 스프링 기본으로 롤백

Dae-Hwa committed:
[REFACTOR] : resource 구분을 하지 않아 필요없어진 프로필 확인 제거

Dae-Hwa committed:
[REFACTOR] : db 접속 정보 소스에서 제거

java 실행 디렉토리에 dbconfig.properties 넣는 방법 사용하도록 변경

Dae-Hwa committed:
[CHORE] : application.properties 로깅 설정 변경

Jiwon-JJW committed:
[FIX]  : jsonToObject 내부에서 JsonFilePath를 입력하기 위한 상수 적

jsonToObject에 있는 ClassPathResource에 상수를 사용하는것으로 변경하였다.

Jiwon-JJW committed:
[FIX] : ObjectMapper final 상수로 변경

ObjectMapper를 상수로 변경해주었다.

Jiwon-JJW committed:
[ADD] : 반찬 Detail 목록에 대한 test case 추가

Test Case에 Detail 모든 항목을 추가했다.

Jiwon-JJW committed:
[FIX] : test case MethodSource 사용하는 것으로 변경

Test Case에 Detail을 assertAll에서 MethodSource를 사용하는것으로 변경하였다.

Dae-Hwa committed:
[FIX] : db 외부설정파일과 연결

prod환경에서 java 실행 위치에 dbconfig.properties파일 존재해야 실행 가능

Dae-Hwa committed:
Merge pull request #22 from Jiwon-JJW/BE/SampleDataFactoryRefactoring

[BE] 리팩토링 진행

Dae-Hwa committed:
[REFACTOR] : SampleDataFactory에서 사용하는 예외들의 패키지 이동

같은 패키지로 묶어줌

Dae-Hwa committed:
[REFACTOR] : 불필요한 sql 삭제

Jiwon-JJW committed:
Merge pull request #23 from Dae-Hwa/BE/apply-review/freddie

[BE]리뷰반영(프레디)

Dae-Hwa commented:
> 안녕하세요! 프레디 & Starve!! 리뷰어 Dion입니다.
> Mock API를 정성스럽게 작성해주셨는데, 아쉬운 부분이 많네요.
> 저도 작년에 비슷한 프로젝트를 진행했던 만큼 제가 프로젝트를 진행했던 경험을 바탕으로 들었던 아쉬운 점을 적어볼게요.
> 
> * 데이터를 왜 그대로 가져와서 그대로 사용하는 형태로 정의해주셨나요? 좋지 않은 네이밍이 존재하는 형태인데, 그대로 사용해주신 부분이 의아했습니다.
> * Mock API 작성에만 굉장히 많은 시간이 들어간 것 같은데요, 저라면 간단하게 구현 후, 실제 API를 빠르게 구현시작하는 형태로 진행했을 것 같습니다. 실제로도 그렇게 했고요.
> * 사실 데이터만 전달하는게 끝이 아니라, API 설계가 들어갔으면 했는데, 아쉽네요.
> 
> 이건 제가 단순히 아쉬웠던 부분이니, 그대로 진행해주시면 됩니다.
> 
> 아래의 코멘트 참고해주시고, 궁금한 점이나, 도움이 필요한 부분, 이해가 되지 않는 부분, 토론하고 싶은 부분이 있으면 언제든 코멘트 남겨주세요. 고생하셨습니다~

https://github.com/codesquad-members-2021/sidedish/pull/22#discussion_r619211037
https://github.com/codesquad-members-2021/sidedish/pull/22#discussion_r619215309
https://github.com/codesquad-members-2021/sidedish/pull/22#discussion_r619215524

피드백 주신 부분 전체적으로 API설계에 대한 아쉬움을 말씀해주신 것 같아 따로 말씀드리자면,
기본적으로 API를 제공해주는걸 기준으로 개발하기로 협의가 됐습니다. 

말씀해주신대로 좋지 않아보이는 부분들이 있었지만, 그런 부분을 그대로 구현하고 개선하는 방향이 이번 미션의 목표가 아닐까 하는 생각으로 이렇게 진행해봤습니다. 요구사항을 바탕으로 설계해보는 것을 지난 미션에서 수행해봤기 때문에 그런 생각을 해본 것 같습니다.

서버 세팅에 시간이 오래 걸리고 목업 api작성에는 큰 시간이 들지 않았다고 생각하는데, 그래도 말씀해주신 것처럼 적지 않은 시간을 투자한건 맞습니다. 굳이 이렇게 한 이유는 샘플데이터와 기본 테스트케이스 등에 대한 고민을 하지 않아도 될 것이라 생각하였기때문입니다.

아무래도 시간이 많이 촉박하게 느껴지다보니 꼭 해보고 싶은 것들 위주로 학습하며 프로젝트를 진행하다보니 신경쓰지 못했던 부분들이 많이 나온 것 같습니다. 이제 데이터 모델링이 끝났고 구현 단계인데, 다음 pr에서는 더 좋은 설계를 보여드릴 수 있도록 해보겠습니다 감사합니다!

ksundong commented on .gitignore:
```diff
@@ -0,0 +1,4 @@
+# BE
+.idea
+logs/
+dbconfig.properties
```
👍

ksundong commented on be/sidedish-be/src/main/java/com/codesquad/sidedish/utils/NotParsingJsonFileException.java:
```diff
@@ -0,0 +1,7 @@
+package com.codesquad.sidedish.utils;
+
+public class NotParsingJsonFileException extends RuntimeException {
```
JsonParsingException 정도면 괜찮을 것 같아요.

ksundong commented on be/sidedish-be/src/main/java/com/codesquad/sidedish/utils/NotParsingJsonFileException.java:
```diff
@@ -0,0 +1,7 @@
+
+public class NotParsingJsonFileException extends RuntimeException {
+    public NotParsingJsonFileException(String jsonFilePath, Throwable e) {
+        super("Json파일 파싱 중 에러 발생. jsonFilePath: " + jsonFilePath, e);
```
하드코드는 상수로 빼는 습관을 들여주세요.
예외 포장을 할 수 있도록 정의해주신 것은 좋네요.

ksundong commented on be/sidedish-be/src/main/java/com/codesquad/sidedish/utils/NotReadJsonFileException.java:
```diff
@@ -0,0 +1,7 @@
+package com.codesquad.sidedish.utils;
```
패키지는 다른 곳으로 이동시켜주시는게 좋을 것 같아요.

ksundong commented on be/sidedish-be/src/main/java/com/codesquad/sidedish/utils/SampleDataFactory.java:
```diff
@@ -0,0 +1,70 @@
+    private static final String JSON_PATH_PREFIX = "sample-data/";
+    private static final String JSON_EXTENSION = ".json";
+
+    public static final ObjectMapper objectMapper = new ObjectMapper()
+            .setPropertyNamingStrategy(PropertyNamingStrategy.SNAKE_CASE);
```
Bean으로 등록해봅시다.
ObjectMapper를 만들어내는 Factory도 있을거에요.
한 번 찾아보세요~

ksundong commented on be/sidedish-be/src/main/java/com/codesquad/sidedish/utils/SampleDataFactory.java:
```diff
@@ -0,0 +1,70 @@
+    public static final ObjectMapper objectMapper = new ObjectMapper()
+            .setPropertyNamingStrategy(PropertyNamingStrategy.SNAKE_CASE);
+
+    private SampleDataFactory() {
+    }
```
💯

ksundong commented on be/sidedish-be/src/main/java/com/codesquad/sidedish/utils/SampleDataFactory.java:
```diff
@@ -0,0 +1,70 @@
+    }
+
+    public static List<SidedishDTO> createBestSidedishes() {
+        return jsonToObject("best-sidedishes", new TypeReference<List<SidedishDTO>>() {
```
이런 형태로 구현해주신 이유가 있나요? 조금 일반적이진 않은 것 같네요 ㅎㅎ

ksundong replied:
여기 존재하는 파일명도 상수로 추출해봅시다.

ksundong commented on be/sidedish-be/src/main/java/com/codesquad/sidedish/utils/SampleDataFactory.java:
```diff
@@ -0,0 +1,70 @@
+        List<DetailDTO> detailDTOs = jsonToObject("details", new TypeReference<List<DetailDTO>>() {
+        });
+
+        return detailDTOs.stream().collect(Collectors.toMap(DetailDTO::getHash, detailDTO -> detailDTO));
```
Stream의 사용 좋습니다.

ksundong commented on be/sidedish-be/src/main/java/com/codesquad/sidedish/utils/SampleDataFactory.java:
```diff
@@ -0,0 +1,70 @@
+    }
+
+    private static <E> E jsonToObject(String jsonFileName, TypeReference<E> typeReference) {
+        try (InputStream is = new ClassPathResource(JSON_PATH_PREFIX + jsonFileName + JSON_EXTENSION).getInputStream()) {
+            return objectMapper.readValue(is, typeReference);
+        } catch (JsonParseException e) {
+            throw new NotParsingJsonFileException(JSON_PATH_PREFIX + jsonFileName + JSON_EXTENSION, e);
+        } catch (IOException e) {
+            throw new NotReadJsonFileException(JSON_PATH_PREFIX + jsonFileName + JSON_EXTENSION, e);
```
`JSON_PATH_PREFIX + jsonFileName + JSON_EXTENSION`
중복되죠?

ksundong commented on be/sidedish-be/src/main/java/com/codesquad/sidedish/web/sidedish/DTO/DetailDTO.java:
```diff
@@ -0,0 +1,61 @@
+    }
+
+    public static DetailDTOBuilder builder() {
+        return DetailDTOBuilder.create();
```
생성자를 바로 사용하면 안될까요?

ksundong commented on be/sidedish-be/src/main/java/com/codesquad/sidedish/web/sidedish/DTO/DetailDataDTO.java:
```diff
@@ -0,0 +1,164 @@
+    }
+
+    public String getDeliveryFee() {
+        return deliveryFee.getFormattedPrice() + " (40,000원 이상 구매 시 무료)";
```
상수는 추출해줍시다.

ksundong commented on be/sidedish-be/src/main/java/com/codesquad/sidedish/web/sidedish/DTO/SidedishDTO.java:
```diff
@@ -0,0 +1,79 @@
+public class SidedishDTO {
+    private long categoryId;
+    private String name;
+    private List<ItemDTO> items = Collections.emptyList();
```
Collections.emptyList()를 사용하면 어떤 List가 생성되나요?

ksundong replied:
의도하신게 맞나요?

ksundong replied:
https://www.baeldung.com/java-collections-emptylist-new-list

ksundong commented on be/sidedish-be/src/main/java/com/codesquad/sidedish/web/sidedish/Price.java:
```diff
@@ -0,0 +1,33 @@
+        try {
+            return NumberFormat.getInstance().parse(price).longValue();
+        } catch (ParseException e) {
+            throw new IllegalArgumentException("Error occurred while parsing price, price : " + price, e);
```
여기도 Custom 예외를 적용할 수 있겠죠?

ksundong commented on be/sidedish-be/src/main/java/com/codesquad/sidedish/web/sidedish/Price.java:
```diff
@@ -0,0 +1,33 @@
+public class Price {
+    private static final String POSTFIX = "원";
+
+    private final long price;
```
값 객체라면 Equals HashCode도 재정의 해주시길 권장드리겠습니다.

ksundong commented on be/sidedish-be/src/main/resources/application-default.properties:
```diff
@@ -0,0 +1,13 @@
+spring.config.import=classpath:common.properties
```
-default를 빼면 default로 동작합니다.

ksundong commented on be/sidedish-be/src/test/java/com/codesquad/sidedish/utils/SampleDataFactoryTest.java:
```diff
@@ -0,0 +1,128 @@
+
+class SampleDataFactoryTest {
+
+    ObjectMapper objectMapper = new ObjectMapper().setPropertyNamingStrategy(PropertyNamingStrategy.SNAKE_CASE);
```
테스트를 위해서 만들어주신거죠?
사실 이런건 매 테스트 마다 재생성될 필요가 없으니 static final을 사용해 정의해주셔도 좋을 듯 합니다.

ksundong replied:
현재도 재생성되진 않지만, 변경가능하고, 로딩 시점의 차이가 있습니다.

ksundong reviewed (approved):
고생하셨습니다.
얼마남지 않은 기간동안 많이 발전하셨으면 좋겠네요.
많은 팀들이 제공된 API를 기준으로 개발을 많이 하셨는데, 그 점이 많이 아쉬웠어요.

작년 제가 했던 팀의 repo입니다. https://github.com/codesquad-member-2020/sidedish-03
저희 팀의 문서를 한 번 보시면 좋을 것 같아요. https://github.com/codesquad-member-2020/sidedish-03/wiki/API-Document
그 때 당시에는 여러분들과 같이 REST API에 대한 개념이 부족했어서 Path는 조금 아쉽네요.

항상 최선을 지향하는 형태로 개발해주시는 습관을 들여주시는게 좋습니다. 기존의 관습에 도전해보도록 해봅시다.
요구사항을 바탕으로 설계하는 것은 당연하지만, 요구사항은 언제든 변경가능하다는 점도 생각해보세요!

목업 API는 클라이언트 개발자의 빠른 개발을 위해서도 필요하지만, 목업 API를 토대로 설계의 잘못된 점을 파악하고 수정할 수 있다는 점입니다. 너무 클라이언트 개발자와 약속한 내용이 절대적인 약속이라는 생각을 안하는게 좋을 것 같아요.
사람인 이상 설계가 바뀌고 더 좋은 설계가 생각날 수 있습니다. 다만, Breaking Change를 피할 수 있게끔 해주는 것이 더 중요하겠죠.

이것 역시 정답은 없으니, 해보고 싶은 대로 해보시는 것도 방법이라고 생각됩니다. 고생하셨어요 ㅎㅎ

시간이 촉박하기에 목업에 너무 많은 시간을 사용하신 것 같다는 생각이었습니다. 다음 설계 기대해보겠습니다! 고생하셨습니다~
